### PR TITLE
lsfd: use extra information loaded from /proc/net/netlink

### PR DESCRIPTION
**Pull request by masatake**
Example output:
```
% sudo ./lsfd -Q '(TYPE == "NETLINK")' | head          
COMMAND             PID            USER ASSOC MODE    TYPE SOURCE MNTID    INODE NAME
systemd               1            root    12  rw- NETLINK sockfs     9    19032 protocol=audit lport=2212937619
systemd               1            root    19  rw- NETLINK sockfs     9      924 protocol=kobject_uevent lport=1 groups=2
systemd               1            root   505  rw- NETLINK sockfs     9    19486 protocol=audit lport=1 groups=1
systemd               1            root   532  rw- NETLINK sockfs     9    30837 protocol=kobject_uevent lport=2495777931 groups=1
systemd-journal     758            root     6  rw- NETLINK sockfs     9    19486 protocol=audit lport=1 groups=1
systemd-udevd       780            root     3  rw- NETLINK sockfs     9    30837 protocol=kobject_uevent lport=2495777931 groups=1
systemd-resolve     981 systemd-resolve     8  rw- NETLINK sockfs     9    20759 protocol=route lport=981 groups=273
auditd              984            root     3  rw- NETLINK sockfs     9    31173 protocol=audit lport=984
dbus-broker        1024            dbus     3  rw- NETLINK sockfs     9    37183 protocol=audit lport=1024
```

### CODE REVIEW

**Comment by masatake** [comment]
This one will be the last pull request I send before v2.39.

**Comment by masatake** [comment]
`hardlink/options` test case failed. I wonder why.

**Comment by t-8ch** [comment]
> `hardlink/options` test case failed. I wonder why.

It has been broken for a long time in the distcheck build. No clue why. Just ignore it.

**Comment by masatake** [comment]
Oh, I see. Thank you.

File: tests/ts/lsfd/mkfds-netlink [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	    printf "LPORT: FAILED (PID: %d != NETLINK.LPORT: %d)\n" "${PID}" "${NETLINK_LPORT}"
+	fi
+	kill -CONT "${PID}"
+    fi
+    wait "${MKFDS_PID}"
+} > "$TS_OUTPUT" 2>&1
+ts_finalize_subtest
+
+ts_init_subtest "groups"
+if ts_skip_nonroot_subtest; then
```
[the remark this answers is not in the document]
Comment (masatake) [reply tests/ts/lsfd/mkfds-netlink]: I split the test case into two.

File: misc-utils/lsfd-sock-xinfo.c [hunk 2]
```diff
@@ -1378,3 +1381,212 @@ static void load_xinfo_from_proc_icmp6(ino_t netns_inode)
...
+	struct sock_xinfo sock;
+	uint16_t protocol;
+	uint32_t lportid;	/* netlink_diag may provide rportid.  */
+	uint32_t groups;
+};
+
+static const char *netlink_decode_protocol(uint16_t protocol)
+{
+	switch (protocol) {
+#ifdef NETLINK_ROUTE
```
[the remark this answers is not in the document]
Comment (masatake) [reply misc-utils/lsfd-sock-xinfo.c]: I deleted the ifdefs/endifs other than SMC.

File: misc-utils/lsfd-sock-xinfo.c [hunk 3]
```diff
@@ -1378,3 +1381,212 @@ static void load_xinfo_from_proc_icmp6(ino_t netns_inode)
...
+		return true;
+	}
+
+	return false;
+}
+
+static const struct sock_xinfo_class netlink_xinfo_class = {
+	.get_name = netlink_get_name,
+	.get_type = netlink_get_type,
+	.get_state = NULL,
```
[the remark this answers is not in the document]
Comment (masatake) [reply misc-utils/lsfd-sock-xinfo.c:1501]: I wonder what I should do.
In this context, `.get_state = NULL` represents that the author of the code considered the necessity of the method for netlink,  and the person concluded that netlink doesn't need such a method. So keeping `= NULL` is not so bad.

[the remark this answers is not in the document]
Comment (t-8ch) [reply misc-utils/lsfd-sock-xinfo.c:1501]: It's the same either way.
static variables are zeroed by default, so the explicit `= NULL` doesn't do anything.
If you prefer it for clarity, fine for me.

**Comment by t-8ch** [comment]
On s390x qemu:
```
2023-02-28T18:43:13.9046180Z          lsfd: NETLINK sockets                ...
2023-02-28T18:43:13.9051232Z diff-{{{
2023-02-28T18:43:13.9262380Z --- /home/runner/work/util-linux/util-linux/tests/expected/lsfd/mkfds-netlink-groups	2023-02-28 17:54:17.749516241 +0000
2023-02-28T18:43:13.9263119Z +++ /home/runner/work/util-linux/util-linux/tests/output/lsfd/mkfds-netlink-groups	2023-02-28 18:43:13.823795326 +0000
2023-02-28T18:43:13.9263504Z @@ -1,3 +1 @@
2023-02-28T18:43:13.9263961Z -    3 NETLINK protocol=sock_diag lport= groups=5                  raw        sock_diag              5
2023-02-28T18:43:13.9264594Z -ASSOC,TYPE,NAME,SOCK.STATE,SOCK.TYPE,NETLINK.PROTOCOL,NETLINK.GROUPS: 0
2023-02-28T18:43:13.9265038Z -LPORT: OK
2023-02-28T18:43:13.9265330Z +test_mkfds: failed in socket(): Protocol not supported
2023-02-28T18:43:13.9290436Z }}}-diff
2023-02-28T18:43:13.9290602Z 
2023-02-28T18:43:13.9809594Z  FAILED (lsfd/mkfds-netlink-groups)
2023-02-28T18:43:14.9676004Z          lsfd: NETLINK sockets                ...
2023-02-28T18:43:14.9682453Z diff-{{{
2023-02-28T18:43:14.9874771Z --- /home/runner/work/util-linux/util-linux/tests/expected/lsfd/mkfds-netlink-protocol	2023-02-28 17:54:17.749516241 +0000
2023-02-28T18:43:14.9875593Z +++ /home/runner/work/util-linux/util-linux/tests/output/lsfd/mkfds-netlink-protocol	2023-02-28 18:43:14.887808943 +0000
2023-02-28T18:43:14.9875952Z @@ -1,3 +1 @@
2023-02-28T18:43:14.9876337Z -    3 NETLINK protocol=sock_diag lport=                  raw        sock_diag              0
2023-02-28T18:43:14.9876787Z -ASSOC,TYPE,NAME,SOCK.STATE,SOCK.TYPE,NETLINK.PROTOCOL: 0
2023-02-28T18:43:14.9877125Z -LPORT: OK
2023-02-28T18:43:14.9877494Z +test_mkfds: failed in socket(): Protocol not supported
2023-02-28T18:43:14.9900666Z }}}-diff
2023-02-28T18:43:14.9900794Z 
2023-02-28T18:43:15.0406363Z  FAILED (lsfd/mkfds-netlink-protocol)
```

**Comment by masatake** [comment]
I added ts_skip to the netlink test cases.

**Pull request merged**